performance/vip-tweaks.php

```diff
@@ -80,3 +80,77 @@ function wpcom_vip_media_library_months_with_files() {
 		return $months;
 	}
 }
+
+/**
+ * Improves performance of all the post listing pages that display the months dropdown filter.
+ * A unique transient is created for each post type.
+ */
+if ( is_admin() ) {
+
+	add_filter( 'pre_months_dropdown_query', 'wpcom_vip_pre_months_dropdown_query', 10, 2 );
+	function wpcom_vip_pre_months_dropdown_query( $months, $post_type ) {
+		global $wpdb;
+		
+		// Don't set a transient for trashed post.
+		if ( isset( $_GET['post_status'] ) && "trash" === $_GET['post_status'] ) {
+			return __return_null();
+		}
+
+		$months = get_transient( 'wpcom_vip_pre_months_dropdown_query_'.$post_type );
+
+		if ( false === $months ) {
+
+			$extra_checks = "AND post_status != 'auto-draft'";
+			if ( ! isset( $_GET['post_status'] ) || 'trash' !== $_GET['post_status'] ) {
+				$extra_checks .= " AND post_status != 'trash'";
+			} elseif ( isset( $_GET['post_status'] ) ) {
+				$extra_checks = $wpdb->prepare( ' AND post_status = %s', $_GET['post_status'] );
+			}
+
+			$months = $wpdb->get_results(
+				$wpdb->prepare(
+					"
+					SELECT DISTINCT YEAR( post_date ) AS year, MONTH( post_date ) AS month
+					FROM $wpdb->posts
+					WHERE post_type = %s
+					$extra_checks
+					ORDER BY post_date DESC
+					",
+					$post_type
+				)
+			);
+			set_transient( 'wpcom_vip_pre_months_dropdown_query_'.$post_type, $months );
+		}
+		return $months;
+	}
+}
+/**
+ * Bust the months dropdown cache when needed.
```

Comment:
I'm wondering if it's more efficient to regenerate the list via cron, or even amend the list and update the transient in this action, rather than completely invalidate it... if a site has a lot of editors, they may find they still have periodic long waits for this dropdown.